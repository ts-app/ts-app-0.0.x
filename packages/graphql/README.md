# ts-app/graphql

- @Resolver decorator expose regular functions as GraphQL resolvers.
- Create NPM packages that expose executable GraphQL schema.


```
#  List schema definitions in package
ts-app-graphql -p @ts-app/security

# Start GraphQL server with security schema and MongoDB connection
ts-app-graphql -p @ts-app/security -s securitySchemaDefinition -m mongodb://localhost:27017
```


* GraphQL
  * Create resolvers by decorating functions with @Resolver()
  * Flexible resolver naming and parameter mapping
  * xx
  * Security
      * Security service to manage user and roles
      * JWT Authentication
      * Role based Authorization
  * Working with multiple schemas
    * Merge schema definitions
    * Predefined schema definition for common scalar types
  * Define schema definitions that can be easi
  *  (schema stitching)
