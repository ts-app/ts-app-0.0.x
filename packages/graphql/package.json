{
  "name": "@ts-app/graphql",
  "version": "0.0.4",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "license": "MIT",
  "scripts": {
    "clean": "rm -fr ./dist",
    "compile": "tsc -p config/tsconfig-compile.json && cp src/*.graphqls dist",
    "compile-watch": "tsc -p config/tsconfig-compile.json --watch",
    "coverage": "jest --config config/jest.config.js --coverage --collectCoverage",
    "doc": "typedoc --gitRevision master --mode file --out dist/docs src",
    "lint": "tslint --project config/tsconfig-test.json",
    "pack": "npm pack",
    "prepublishOnly": "npm run lint && npm run compile && npm test",
    "test": "jest --config config/jest.config.js",
    "test-watch": "jest --config config/jest.config.js --watch"
  },
  "bin": {
    "ts-app-graphql": "dist/graphql.js"
  },
  "publishConfig": {
    "access": "public"
  },
  "devDependencies": {
    "apollo-codegen": "^0.17.1",
    "jest": "^21.2.1",
    "tslint": "^5.8.0",
    "typedoc": "^0.9.0",
    "typescript": "~2.5"
  },
  "dependencies": {
    "@ts-app/common": "^0.0.3",
    "@ts-app/mongo": "^0.0.3",
    "body-parser": "^1.18.2",
    "chalk": "^2.3.0",
    "clear-module": "^2.1.0",
    "commander": "^2.11.0",
    "cors": "^2.8.4",
    "express": "^4.16.2",
    "graphql-server-express": "^1.2.0",
    "graphql-tag": "^2.5.0",
    "graphql-tools": "^2.5.1",
    "graphql-voyager": "^1.0.0-rc.9",
    "import-fresh": "^2.0.0",
    "passport": "^0.4.0",
    "passport-jwt": "^3.0.1"
  },
  "peerDependencies": {
    "graphql": ">= 0.10 < 1",
    "react": ">= 14",
    "react-dom": ">= 14"
  }
}
