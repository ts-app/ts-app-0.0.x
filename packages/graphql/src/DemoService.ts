import { Resolver } from './Resolver'
import { ServiceInfo } from './ServiceInfo'

export class DemoService {
  private anotherNumber: number = 123

  constructor (public count: number = 0) {
  }

  @Resolver({ paramNames: [ 'name', 'message' ] })
  echo (name: string, message: string) {
    return `Hello, ${name}! Message is "${message}"`
  }

  @Resolver({ type: 'mutation' })
  add ({ howMany = 1 }: { howMany?: number }) {
    this.count += howMany
    return this.count
  }

  @Resolver({ type: 'mutation' })
  async promiseToSubtract ({ howMany = 1 }: { howMany?: number }): Promise<number> {
    return new Promise<number>(resolve => {
      setTimeout(() => {
        this.count -= howMany
        resolve(this.count)
      }, 1)
    })
  }

  justUseAnotherNumber () {
    return this.anotherNumber * 3
  }

  @Resolver()
  throwSomeError () {
    throw new Error('simulated error')
  }
}

export class DemoServiceWithServiceInfo implements ServiceInfo {
  info (): { serviceName: string } {
    return {
      serviceName: 'CustomServiceName'
    }
  }

  @Resolver({ paramNames: [ 'name', 'message' ] })
  echo2 (name: string, message: string) {
    return `Hello, ${name}! Message is "${message}" - from echo2`
  }
}
