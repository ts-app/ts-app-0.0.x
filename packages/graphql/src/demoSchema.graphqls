type Query {
  echo (name: String!, message: String!): String
  throwSomeError: String
}

type Mutation {
  add (howMany: Int): Int
  promiseToSubtract (howMany: Int): Int
}
