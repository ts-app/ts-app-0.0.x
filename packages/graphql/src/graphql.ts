#! /usr/bin/env node
import * as fs from 'fs'
import * as clearModule from 'clear-module'
import * as commander from 'commander'
import * as importFresh from 'import-fresh'
import { MongoService } from '@ts-app/mongo'
import { makeExecutableSchema } from 'graphql-tools'
import chalk from 'chalk'
import { mergeSchemaDefinitions } from './mergeSchemaDefinitions'
import { serveSchema } from './serveSchema'
import { Server } from 'http'
import { clearWatchMatchActions, watchMatchAction } from './watchMatchAction'
import { SchemaInfo } from './SchemaInfo'
import { printQueryInfo } from './printQueryInfo'

const name = 'ts-app-graphql'
const version = require('../package.json').version
let restartCount = 0

commander.name(name).version(version)
  .option('-l --list')
  .option('-p --package [packageName]')
  .option('-s --schema [schemaDefinitionName]')
  .option('-m --mongoUrl [mongoUrl]', 'Pass an instance of MongoService with specified MongoURL to SchemaDefinition (e.g. mongodb://localhost:27017)')
  .option('-r --rootType [rootType]', 'Diagram root type (e.g. Query, Mutation) Default: "Query".')
  .option('-d --develop', 'Development mode')
  .parse(process.argv)

const log = console.log
const debug = (message: string) => {
  console.log(chalk.green('DEBUG: ') + message)
}

let watchToRestartWarned = false
const watchToRestart = async (dir: string, server?: Server) => {
  if (!fs.existsSync(dir)) {
    if (!watchToRestartWarned) {
      debug(`Directory [${chalk.dim(dir)}] does not exist. Watch for changes disabled`)
      watchToRestartWarned = true
    }
    return
  }

  // watch for changes in "dir", restart server
  watchMatchAction({
    dir, watchOnce: true,
    match: filename => filename.endsWith('.graphqls') || filename.endsWith('.js'),
    action: async filename => {
      const restartStartTime = Date.now()
      if (server) {
        await new Promise(resolve => server.close(resolve))
      }

      const name = `${dir}/${filename}`
      debug(`Reloading ${chalk.dim(name)}`)
      clearModule(name)
      // TODO: removal of @Resolver() from methods not reflected. Because ResolverService cache stuff?

      // tslint:disable-next-line
      serve(false)
        .then(({ error, schemaInfo }) => {
          if (error) {
            log(error)
          } else {
            printQueryInfo(schemaInfo!)
            const restartElapsed = Date.now() - restartStartTime
            restartCount += 1
            if (restartCount === 1) {
              log(chalk.green('INFO: ') + `Server restarted [${chalk.dim(`${restartElapsed}ms`)}]`)
            } else {
              log(chalk.green('INFO: ') + `Server restarted [${chalk.dim(`${restartElapsed}ms`)}] [${chalk.bold(restartCount.toString())} times]`)
            }
          }
        })
        .catch(e => {
          log(chalk.red('ERROR: ') + 'Restarting server')
          log(e)
        })
    }
  })
}

let watchToDevelopWarned = false
const watchToDevelop = (dir: string) => {
  const srcDir = `${dir}/src`
  const distDir = `${dir}/dist`

  if (!(fs.existsSync(srcDir) && fs.existsSync(distDir))) {
    if (!watchToDevelopWarned) {
      debug(`Directories [${chalk.dim(`${srcDir}, ${distDir}`)}] does not exist. Development mode disabled`)
      watchToDevelopWarned = true
    }
    return
  }

  // watch for changes to .graphqls and copy to dist
  watchMatchAction({
    dir: srcDir,
    match: filename => filename.endsWith('.graphqls'),
    action: (filename) => {
      fs.copyFile(`${srcDir}/${filename}`, `${distDir}/${filename}`, err => {
        if (err) {
          log(chalk.red('ERROR: ') + `Copying ${filename}`)
        } else {
          debug(`GraphQL schema copied to /dist [${chalk.dim(`${srcDir}/${filename}`)}]`)
        }
      })
    }
  })
}

const serve = async (showInfo: boolean = true): Promise<{ error?: string, schemaInfo?: SchemaInfo }> => {
  const list = !!commander.list
  let packageName = typeof commander.package === 'string' ? commander.package : null

  if (!packageName) {
    return { error: chalk.red('ERROR: ') + 'Package name must be specified' }
  }

  if (packageName.endsWith('.js') && !packageName.startsWith('/')) {
    // use absolute path when .js file specified with relative path
    packageName = `${process.cwd()}/${packageName}`
  }

  if (!packageName.endsWith('.js') && packageName.indexOf('/') !== -1 && !packageName.startsWith('@')) {
    // for non .js files, auto prefix '@' if need
    packageName = `@${packageName}`
  }

  const pkg = importFresh(packageName)
  const names = Object.keys(pkg)
  const sdNames: any = names.filter(n => n.endsWith('SchemaDefinition'))

  if (list) {
    log(`${chalk.bold(packageName)} has ${chalk.bold(sdNames.length)} schema definition(s)`)
    log(`[${chalk.dim(sdNames)}]`)
    return {}
  }

  if (sdNames.length === 0) {
    return { error: chalk.red('ERROR: ') + `Package [${chalk.dim(packageName)}] has no schema definition` }
  }

  let schemaDefinitionName = typeof commander.schema === 'string' ? commander.schema : null

  if (schemaDefinitionName === null && sdNames.length === 1) {
    schemaDefinitionName = sdNames[ 0 ]
    showInfo && log(chalk.green('AUTO-DETECT: ') + `Using schema definition [${chalk.dim(schemaDefinitionName)}] in package [${chalk.dim(packageName)}]`)
  } else {
    if (sdNames.indexOf(schemaDefinitionName) === -1) {
      return { error: chalk.red('ERROR: ') + `Schema definition [${chalk.dim(schemaDefinitionName)}] not found in package [${chalk.dim(packageName)}]` }
    }
  }

  const sdf = pkg[ schemaDefinitionName ]
  if (typeof sdf !== 'function') {
    return { error: chalk.red('ERROR: ') + `Expected module [${chalk.dim(schemaDefinitionName)}] to be a function` }
  }

  const params: any = {}
  const mongoUrl = typeof commander.mongoUrl === 'string' ? commander.mongoUrl : null
  if (mongoUrl) {
    params[ 'mongoService' ] = new MongoService(mongoUrl)
    log(chalk.green('INFO ') + `MongoService [${chalk.dim(mongoUrl)}] passed to schema definition`)
  }

  const definition = sdf(params)
  const merged = mergeSchemaDefinitions([ definition ])
  const rootType = typeof commander.rootType === 'string' ? commander.rootType : undefined

  // prepare to retry failed server startup on file changes
  clearWatchMatchActions()
  const watchDev = typeof commander.develop === 'boolean'
  if (watchDev) {
    watchToDevelop(process.cwd())
  }
  const watchDir = `${process.cwd()}/dist`
  await watchToRestart(watchDir)

  const schema = makeExecutableSchema(merged)
  const server = await serveSchema({ schema, rootType, showInfo })
  // clear all (previous) watchers and watch again, with server specified
  clearWatchMatchActions()
  if (watchDev) {
    watchToDevelop(process.cwd())
  }
  await watchToRestart(watchDir, server)

  return { schemaInfo: { ...merged, schema } }
}

const startTime = Date.now()
serve(true)
  .then(({ error, schemaInfo }) => {
    if (error) {
      log(error)
    } else if (schemaInfo) {
      const elapsed = Date.now() - startTime
      printQueryInfo(schemaInfo!)
      log(chalk.green('INFO: ') + `Server started [${chalk.dim(`${elapsed}ms`)}]`)
    }
  })
  .catch(e => {
    log(chalk.red('ERROR: ') + 'Starting server')
    log(e)
  })
