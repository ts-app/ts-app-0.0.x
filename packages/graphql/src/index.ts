export * from './demoSchemaDefinition'
export * from './DemoService'
export * from './environmentVariableIsSet'
export * from './mergeSchemaDefinitions'
export * from './mergeTypeDefs'
export * from './Resolver'
export * from './ResolverError'
export * from './ResolverMiddlewareInput'
export * from './ResolverService'
export * from './SchemaDefinition'
export * from './SecurityMiddleware'
export * from './serveSchema'
export * from './ServiceInfo'
export * from './standardSchemaDefinition'
export * from './userIdMatchParam'
