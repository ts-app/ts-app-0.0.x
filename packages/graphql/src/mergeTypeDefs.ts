import { DocumentNode, ObjectTypeDefinitionNode, parse } from 'graphql'

const getInsertLocation = (typeDef: string, rootToInsert: 'Query' | 'Mutation' | string) => {
  const parsed = parse(typeDef)
  const root = parsed.definitions.find(d => {
    if (d.kind === 'ObjectTypeDefinition') {
      const otd: ObjectTypeDefinitionNode = d
      if (otd.name.kind === 'Name' && otd.name.value === rootToInsert && otd.fields.length > 0) {
        return true
      }
    }
    return false
  })

  if (root) {
    const otd = root as ObjectTypeDefinitionNode
    return otd.fields[ otd.fields.length - 1 ].loc!.end
  }

  // rootToInsert does not exist
  return null
}

const extract = (typeDef: string, rootToExtract: 'Query' | 'Mutation', rootsToExclude: string[]) => {
  // --- GraphQL parse type definitions
  const parsed = parse(typeDef)

  let root: ObjectTypeDefinitionNode | null = null
  const excludedLocs: { start: number, end: number }[] = []

  const excludes = Array.from(new Set([ ...rootsToExclude, rootToExtract ]))
  parsed.definitions.map(d => {
    if (d.kind === 'ObjectTypeDefinition') {
      const otd: ObjectTypeDefinitionNode = d

      if (otd.name.kind === 'Name' && otd.name.value === rootToExtract && otd.fields.length > 0) {
        root = otd
      }
      if (otd.name.kind === 'Name' && excludes.indexOf(otd.name.value) !== -1 && otd.loc) {
        excludedLocs.push(otd.loc)
      }
    }
  })

  // --- extract specified root queries
  let queries
  if (root) {
    const otd = root as ObjectTypeDefinitionNode
    const start = otd.fields[ 0 ].loc!.start
    const end = otd.fields[ otd.fields.length - 1 ].loc!.end

    queries = typeDef.substring(start, end)
  } else {
    queries = ''
  }

  // --- extract type definition with unwanted roots filtered
  let reduced = excludedLocs.reduce<{ result: string, startIdx: number }>((prev, loc) => {
    prev.result += typeDef.substring(prev.startIdx, loc.start)
    prev.startIdx = loc.end
    return prev
  }, {
    result: '',
    startIdx: 0
  })
  const filteredTypeDef = reduced.result + typeDef.substring(reduced.startIdx)

  return {
    queries,
    filteredTypeDef
  }
}

const getLoc = (doc: DocumentNode, rootType: string) => {
  const definition = doc.definitions.find(d => {
    if (d.kind === 'ObjectTypeDefinition') {
      return d.name.kind === 'Name' && d.name.value === rootType
    }
    return false
  })

  return definition ? definition.loc : null
}

const removeRootType = (typeDef: string, rootType: string) => {
  const loc = getLoc(parse(typeDef), rootType)
  return loc ? typeDef.substring(0, loc.start) + typeDef.substring(loc.end) : typeDef
}

const mergeRootType = (mainTypeDef: string, typeDefs: string[], rootType: 'Query' | 'Mutation', appendFilteredTypeDef: boolean = false) => {
  const insertAt = getInsertLocation(mainTypeDef, rootType)

  const reduced = typeDefs.reduce<{ queries: string, filteredTypeDef: string }>((prev, typeDef) => {
    const current = extract(typeDef, rootType, [ 'Query', 'Mutation' ])
    prev.queries += current.queries
    prev.filteredTypeDef += current.filteredTypeDef
    return prev
  }, {
    queries: '',
    filteredTypeDef: ''
  })

  let merged = ''

  if (insertAt === null) {
    merged += `type ${rootType} {\n`
  } else {
    merged = mainTypeDef.substring(0, insertAt)
  }
  merged += '\n' + reduced.queries

  if (insertAt === null) {
    merged += `}\n${removeRootType(mainTypeDef, rootType)}`
  } else {
    merged += mainTypeDef.substring(insertAt)
  }

  if (appendFilteredTypeDef) {
    merged += `\n${reduced.filteredTypeDef}`
  }

  return merged
}

/**
 * Merge an array of type definition strings. This function allows "Query" and "Mutation" root types
 * to be repeated - by appending contents of such repeated root types.
 */
export const mergeTypeDefs = (typeDefs: string[]): string => {
  const first = typeDefs.shift() as string

  const mergedQuery = mergeRootType(first, typeDefs, 'Query')

  return mergeRootType(mergedQuery, typeDefs, 'Mutation', true)
}
