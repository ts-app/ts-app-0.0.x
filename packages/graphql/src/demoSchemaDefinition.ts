import { loadFile } from '@ts-app/common'
import { DemoService } from './DemoService'
import { ResolverService } from './ResolverService'
import { standardSchemaDefinition } from './standardSchemaDefinition'
import { SchemaDefinition } from './SchemaDefinition'

export const demoSchemaDefinition = (): SchemaDefinition => {
  const resolver = ResolverService.getInstance()
  resolver.registerService(new DemoService())
  const typeDefs = loadFile(`${__dirname}/demoSchema.graphqls`)
  const resolvers = resolver.makeResolvers()
  const standard = standardSchemaDefinition()

  return {
    typeDefs,
    resolvers,
    dependencies: {
      standard
    }
  }
}
