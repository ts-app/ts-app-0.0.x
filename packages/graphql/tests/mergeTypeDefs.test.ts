import { mergeTypeDefs } from '../src/mergeTypeDefs'
import { parse, buildASTSchema } from 'graphql'

describe('mergeTypeDefs', () => {
  test('no Query/Mutation (only scalar types)', () => {
    const merged = mergeTypeDefs([ 'scalar Date', 'scalar Book' ])
    expect(merged).toMatchSnapshot()
  })

  test('scalar with query', () => {
    const schema1 = `scalar Date`

    const schema2 = `type Query {
      echo(name: String): String
    }`

    const merged = mergeTypeDefs([ schema1, schema2 ])
    expect(merged).toMatchSnapshot()
  })

  test('nothing to merge scenario', () => {
    const schema1 = `type Query {
      echo(name: String): String
    }
    scalar Book
type Mutation {}`

    const schema2 = `scalar Date`

    const merged = mergeTypeDefs([ schema1, schema2 ])
    expect(merged).toMatchSnapshot()
  })

  test('simple merge', () => {
    const schema1 = `type Query {
echo(name: String): String
    }
type Mutation {
reset: Boolean
}`

    const schema2 = `type Query {
hello: String
    }`

    const merged = mergeTypeDefs([ schema1, schema2 ])
    expect(merged).toMatchSnapshot()
  })

  test('typical scenario', () => {
    const schema1 = `scalar Date
type Query {
  echo (name: String!, message: String!): String
}

type Mutation {
  add (howMany: Int): Int
}
`
    const schema2 = `type Query {
    users(q: String): UsersPayload
}

type UsersPayload {
    error: Date
    username: [String]
    # from "schema1"
    sampleDate: Date
}

type Mutation {
    createUser(username: String): Boolean
}

scalar Time`

    const merged = mergeTypeDefs([ schema1, schema2 ])
    expect(merged).toMatchSnapshot()

    // verify merged type definition can be parsed as AST
    const value = parse(merged)
    const ast = buildASTSchema(value)
    expect(ast.getQueryType()).toMatchSnapshot()
    expect(ast.getMutationType()).toMatchSnapshot()
  })
})
