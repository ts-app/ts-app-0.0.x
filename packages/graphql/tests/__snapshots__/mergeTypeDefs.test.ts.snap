// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeTypeDefs no Query/Mutation (only scalar types) 1`] = `
"type Mutation {

}
type Query {

}
scalar Date
scalar Book"
`;

exports[`mergeTypeDefs nothing to merge scenario 1`] = `
"type Mutation {

}
type Query {
      echo(name: String): String

    }
    scalar Book

scalar Date"
`;

exports[`mergeTypeDefs scalar with query 1`] = `
"type Mutation {

}
type Query {

echo(name: String): String}
scalar Date
"
`;

exports[`mergeTypeDefs simple merge 1`] = `
"type Query {
echo(name: String): String
hello: String
    }
type Mutation {
reset: Boolean

}
"
`;

exports[`mergeTypeDefs typical scenario 1`] = `
"scalar Date
type Query {
  echo (name: String!, message: String!): String
users(q: String): UsersPayload
}

type Mutation {
  add (howMany: Int): Int
createUser(username: String): Boolean
}



type UsersPayload {
    error: Date
    username: [String]
    # from \\"schema1\\"
    sampleDate: Date
}



scalar Time"
`;

exports[`mergeTypeDefs typical scenario 2`] = `"Query"`;

exports[`mergeTypeDefs typical scenario 3`] = `"Mutation"`;
